Code/src/main/java/nl/utwente/group10/ui/components/InputAnchor.java

```diff
@@ -1,47 +1,67 @@
 package nl.utwente.group10.ui.components;
 
-import javafx.scene.input.MouseEvent;
 import nl.utwente.group10.haskell.expr.Expr;
 import nl.utwente.group10.haskell.expr.Ident;
 import nl.utwente.group10.ui.CustomUIPane;
+import nl.utwente.group10.ui.handlers.InputAnchorHandler;
 
 import java.io.IOException;
 import java.util.Optional;
 
 public class InputAnchor extends ConnectionAnchor {
-    private Optional<Connection> up;
 
-    public InputAnchor(Block block, CustomUIPane pane) throws IOException {
-        super(block, pane);
+	public InputAnchor(Block block, CustomUIPane pane) throws IOException {
+		super(block, pane);
+		new InputAnchorHandler(pane.getConnectionCreationManager(),this);
+	}
 
-        up = Optional.empty();
+	public Expr asExpr() {
+		if (isConnected()) {
+			return getConnection().get().getOutputAnchor().get().getBlock().asExpr();
+		} else {
+			return new Ident("undefined");
+		}
+	}
 
-        this.addEventHandler(MouseEvent.MOUSE_CLICKED, event -> {
-            up.map(oldAnchor -> {
-                return pane.getChildren().removeAll(oldAnchor);
-            });
-
-            pane.getLastOutputAnchor().map(anchor -> {
-                try {
-                    Connection upstream = new Connection(anchor, this);
-                    pane.getChildren().addAll(upstream);
-                    up = Optional.of(upstream);
-                } catch (IOException e) {
-                    e.printStackTrace();
-                }
-
-                return null;
-            });
-
-            pane.invalidate();
-        });
+	@Override
+	public Optional<Connection> createConnectionWith(ConnectionAnchor other) {
+		if(other instanceof OutputAnchor){
+			return createConnectionFrom((OutputAnchor) other);
+		}else{
+			return Optional.empty();
+		}
+	}
+	
+	public Optional<Connection> createConnectionFrom(OutputAnchor other){
+		if(!isConnected()){
+			new Connection(this, (OutputAnchor) other);
+			getPane().getChildren().add(getConnection().get());
+			getPane().invalidate();
+			return getConnection();
+		}else{
+			return Optional.empty();
+		}
+	}
+	
+	@Override
+	public String toString(){
+    	return "InputAnchor belonging to "+getBlock().getName();
     }
 
-    public Expr asExpr() {
-        if (up.isPresent()) {
-            return up.get().getInputFunction().asExpr();
-        } else {
-            return new Ident("undefined");
-        }
-    }
+	@Override
+	public boolean canConnect() {
+		//InputAnchors only support 1 connection;
+		return !isConnected();
+	}
+
+	@Override
+	public void disconnect(Connection connection) {
+		if(connection.equals(getConnection().get())){
+			setConnection(null);
+		}else{
+			throw new MultipleConnectionsException();
+		}
+	}
+	
+	public class MultipleConnectionsException extends RuntimeException{}
```

Comment:
Weet niet of ik dit echt een exception vind, en het is wat eng dat je 'm hier wel gooit maar nergens vangt...